src/dda/utils/git/changeset.py

```diff
@@ -0,0 +1,254 @@
+    _changes: dict[Path, ChangedFile] = field(default_factory=dict)
+
+    """
+    Represents a set of changes to files in a git repository.
+    This can both be a change between two commits, or the changes in the working directory.
+
+    When considering the changes to the working directory, the untracked files are considered as added files.
+    """
+
+    # == dict proxy methods == #
+    def keys(self) -> dict_keys[Path, ChangedFile]:
+        return self._changes.keys()
+
+    def values(self) -> dict_values[Path, ChangedFile]:
+        return self._changes.values()
+
+    def items(self) -> dict_items[Path, ChangedFile]:
+        return self._changes.items()
+
+    def __getitem__(self, key: Path) -> ChangedFile:
+        return self._changes[key]
+
+    def __contains__(self, key: Path) -> bool:
+        return key in self._changes
+
+    def __len__(self) -> int:
+        return len(self._changes)
+
+    def __iter__(self) -> Iterator[Path]:
+        return iter(self._changes.keys())
+
+    def __or__(self, other: Self) -> Self:
+        return self.from_iter(list(self.values()) + list(other.values()))
+
+    # == properties == #
+    @cached_property
+    def added(self) -> set[Path]:
+        """List of files that were added."""
+        return {change.file for change in self.values() if change.type == ChangeType.ADDED}
+
+    @cached_property
+    def modified(self) -> set[Path]:
+        """List of files that were modified."""
+        return {change.file for change in self.values() if change.type == ChangeType.MODIFIED}
+
+    @cached_property
+    def deleted(self) -> set[Path]:
+        """List of files that were deleted."""
+        return {change.file for change in self.values() if change.type == ChangeType.DELETED}
+
+    @cached_property
+    def changed(self) -> set[Path]:
+        """List of files that were changed (added, modified, or deleted)."""
+        return set(self.keys())
+
+    # == methods == #
+    def digest(self) -> str:
+        """Compute a hash of the changeset."""
+        from hashlib import sha1
+
+        digester = sha1()  # noqa: S324
+        for change in sorted(self.values(), key=lambda x: x.file.as_posix()):
+            digester.update(change.file.as_posix().encode())
+            digester.update(change.type.value.encode())
+            digester.update(change.patch.encode())
+
+        return str(digester.hexdigest())
+
+    @classmethod
+    def from_iter(cls, data: Iterable[ChangedFile]) -> Self:
+        """Create a ChangeSet from an iterable of FileChanges."""
+        items = {change.file: change for change in data}
+        return cls(_changes=items)
+
+    @classmethod
+    def generate_from_diff_output(cls, diff_output: str | list[str]) -> Self:
+        """
+        Generate a changeset from the output of a git diff command.
+        The output should be passed as a string or a list of lines.
+        """
+        return cls.from_iter(ChangedFile.generate_from_diff_output(diff_output))
```

Comment:
I'm not sure when these whatever actually be used. Do you have concrete use cases in mind?